"progress" format does not print failures.

When a test fails, the failure details are not printed:

![image](https://user-images.githubusercontent.com/1386232/80009292-c7dda380-84c0-11ea-93e0-f80a1efa2fad.png)


Let's change this so it re-uses the same summary printing from Documentation format.

Code is here: 
https://github.com/deathowl/TestSlide/tree/print_fail_progress
test fails for some clowny reason (control characters or something)
```
AssertionError: '\x1b[216 chars]1m\r\r\nFAILED EXAMPLES:\x1b[0m\r\n\x1b[0m\x1b[130 chars]\r\n' != '\x1b[216 chars]1m\r\nFAILED EXAMPLES:\x1b[0m\x1b[0m\x1b[31m\r[124 chars]\r\n'
+ .F.SS.F.S
?                                                                                                                                         +
-
- FAILED EXAMPLES:
+ FAILED EXAMPLES:
?                     +++++++++
- 	failing example: SimulatedFailure: test failure (extra)
? ---------
+ 	failing example: SimulatedFailure: test failure (extra)
?                                                             +++++++++
- 	test_failing: AssertionError:
? ---------
+ 	test_failing: AssertionError:
-
 : Command ['/Library/Developer/CommandLineTools/usr/bin/python3', '-m', 'testslide.cli', '--show-testslide-stack-trace', '/Users/death0wl/TestSlide/tests/sample_tests.py', '--format', 'progress'] expected to have have this stdout:

.F.SS.F.S
FAILED EXAMPLES:
	failing example: SimulatedFailure: test failure (extra)
	test_failing: AssertionError:


But output was different:
STDOUT:
.F.SS.F.S
FAILED EXAMPLES:
	failing example: SimulatedFailure: test failure (extra)
	test_failing: AssertionError:


```

--- a/tests/cli_unittest.py
+++ b/tests/cli_unittest.py
@@ -555,15 +555,7 @@ class TestCliProgressFormatter(TestCliBase):
     def test_ouputs_dots(self):
         self.run_testslide(
             expected_return_code=1,
-            expected_stdout=(
-                ".F.SS.F.S"
-                + "\nFAILED EXAMPLES:"
-                + "\n"
-                + "\tfailing example: SimulatedFailure: test failure (extra)"
-                + "\n"
-                + "\ttest_failing: AssertionError: "
-                + "\n\n"
-            ),
+            expected_stdout_startswith=(".F.SS.F.S" + "\nFailures:"),
         )
 
     def test_ouputs_colored_dots_with_terminal(self):
@@ -571,7 +563,7 @@ class TestCliProgressFormatter(TestCliBase):
         self.run_testslide(
             tty_stdout=True,
             expected_return_code=1,
-            expected_stdout=(
+            expected_stdout_startswith=(
                 self.green(".")
                 + self.red("F")
                 + self.green(".")
@@ -581,12 +573,7 @@ class TestCliProgressFormatter(TestCliBase):
                 + self.red("F")
                 + self.green(".")
                 + self.yellow("S")
-                + self.red("\r\nFAILED EXAMPLES:")
-                + "\r\n"
-                + self.red("\tfailing example: SimulatedFailure: test failure (extra)")
-                + "\r\n"
-                + self.red("\ttest_failing: AssertionError: ")
-                + "\r\n\r\n"
+                + self.red("\r\nFailures:")
             ),
         )
 

--- a/tests/cli_unittest.py
+++ b/tests/cli_unittest.py
@@ -102,7 +102,6 @@ class TestCliBase(unittest.TestCase):
             output += f"STDOUT:\n{stdout_output}\n"
         if stderr_output:
             output += f"STDERR:\n{stderr_output}\n"
-
         self.assertEqual(
             return_code,
             expected_return_code,
@@ -554,9 +553,21 @@ class TestCliProgressFormatter(TestCliBase):
         self.argv.append("progress")
 
     def test_ouputs_dots(self):
-        self.run_testslide(expected_return_code=1, expected_stdout=(".F.SS.F.S\n"))
+        self.run_testslide(
+            expected_return_code=1,
+            expected_stdout=(
+                ".F.SS.F.S"
+                + "\nFAILED EXAMPLES:"
+                + "\n"
+                + "\tfailing example: SimulatedFailure: test failure (extra)"
+                + "\n"
+                + "\ttest_failing: AssertionError: "
+                + "\n\n"
+            ),
+        )
 
     def test_ouputs_colored_dots_with_terminal(self):
+        self.maxDiff = None
         self.run_testslide(
             tty_stdout=True,
             expected_return_code=1,
@@ -570,7 +581,12 @@ class TestCliProgressFormatter(TestCliBase):
                 + self.red("F")
                 + self.green(".")
                 + self.yellow("S")
+                + self.red("\r\nFAILED EXAMPLES:")
                 + "\r\n"
+                + self.red("\tfailing example: SimulatedFailure: test failure (extra)")
+                + "\r\n"
+                + self.red("\ttest_failing: AssertionError: ")
+                + "\r\n\r\n"
             ),
         )
 

--- a/tests/cli_unittest.py
+++ b/tests/cli_unittest.py
@@ -255,7 +255,7 @@ class TestCliQuiet(TestCliBase):
         )
 
 
-class TestCliDocumentationFormatter(TestCliBase):
+class TestCliDocumentFormatter(TestCliBase):
     def setUp(self):
         super().setUp()
         self.argv = ["--format", "documentation"] + self.argv

--- a/testslide/runner.py
+++ b/testslide/runner.py
@@ -178,6 +178,7 @@ class ColorFormatterMixin(BaseFormatter):
     def print_cyan(self, *values, **kwargs):
         self._print_attrs("36", *values, **kwargs)
 
+
 class FailurePrinterMixin(ColorFormatterMixin):
     def print_failed_example(self, number, example, exception):
         self.print_white(
@@ -212,6 +213,7 @@ class FailurePrinterMixin(ColorFormatterMixin):
                     '      File "{}", line {}, in {}\n'
                     "        {}".format(path, line, function_name, text)
                 )
+
 
 class SlowImportWarningMixin(ColorFormatterMixin):
     def __init__(self, *args, **kwargs):
@@ -376,7 +378,6 @@ class DocumentFormatter(DSLDebugMixin, SlowImportWarningMixin, FailurePrinterMix
                 skip_text="" if self._color_output() else ": SKIP",
             )
         )
-
 
     def finish(self, not_executed_examples):
         super().finish(not_executed_examples)

--- a/testslide/runner.py
+++ b/testslide/runner.py
@@ -178,6 +178,40 @@ class ColorFormatterMixin(BaseFormatter):
     def print_cyan(self, *values, **kwargs):
         self._print_attrs("36", *values, **kwargs)
 
+class FailurePrinterMixin(ColorFormatterMixin):
+    def print_failed_example(self, number, example, exception):
+        self.print_white(
+            "  {number}) {context}: {example}".format(
+                number=number, context=example.context.full_name, example=example
+            )
+        )
+        if type(exception) is AggregatedExceptions:
+            exception_list = exception.exceptions
+        else:
+            exception_list = [exception]
+        for number, exception in enumerate(exception_list):
+            self.print_red(
+                "    {number}) {exception_class}: {message}".format(
+                    number=number + 1,
+                    exception_class=exception.__class__.__name__,
+                    message="\n    ".join(str(exception).split("\n")),
+                )
+            )
+            for path, line, function_name, text in traceback.extract_tb(
+                exception.__traceback__
+            ):
+                if not self.show_testslide_stack_trace and path.startswith(
+                    os.path.dirname(__file__)
+                ):
+                    continue
+                if self.trim_path_prefix:
+                    split = path.split(self.trim_path_prefix)
+                    if len(split) == 2 and not split[0]:
+                        path = split[1]
+                self.print_cyan(
+                    '      File "{}", line {}, in {}\n'
+                    "        {}".format(path, line, function_name, text)
+                )
 
 class SlowImportWarningMixin(ColorFormatterMixin):
     def __init__(self, *args, **kwargs):
@@ -257,7 +291,7 @@ class QuietFormatter(BaseFormatter):
     pass
 
 
-class ProgressFormatter(DSLDebugMixin, SlowImportWarningMixin, ColorFormatterMixin):
+class ProgressFormatter(DSLDebugMixin, SlowImportWarningMixin, FailurePrinterMixin):
     """
     Simple formatter that outputs "." when an example passes or "F" w
     """
@@ -282,20 +316,16 @@ class ProgressFormatter(DSLDebugMixin, SlowImportWarningMixin, ColorFormatterMix
     def finish(self, not_executed_examples):
         super().finish(not_executed_examples)
         if self.results["fail"] and not self.dsl_debug:
-            self.print_red("\nFAILED EXAMPLES:")
-            for fail in self.results["fail"]:
-                self.print_red(
-                    "\t{example}: {ex_class}: {ex_message}".format(
-                        example=fail["example"],
-                        ex_class=type(fail["exception"]).__name__,
-                        ex_message=str(fail["exception"]).split("\n")[0],
-                    )
+            self.print_red("\nFailures:")
+            for number, result in enumerate(self.results["fail"]):
+                print("")
+                self.print_failed_example(
+                    number + 1, result["example"], result["exception"]
                 )
-
         print("")
 
 
-class DocumentFormatter(DSLDebugMixin, SlowImportWarningMixin, ColorFormatterMixin):
+class DocumentFormatter(DSLDebugMixin, SlowImportWarningMixin, FailurePrinterMixin):
     def get_dsl_debug_indent(self, example):
         return "  " * (example.context.depth + 1)
 
@@ -347,39 +377,6 @@ class DocumentFormatter(DSLDebugMixin, SlowImportWarningMixin, ColorFormatterMix
             )
         )
 
-    def print_failed_example(self, number, example, exception):
-        self.print_white(
-            "  {number}) {context}: {example}".format(
-                number=number, context=example.context.full_name, example=example
-            )
-        )
-        if type(exception) is AggregatedExceptions:
-            exception_list = exception.exceptions
-        else:
-            exception_list = [exception]
-        for number, exception in enumerate(exception_list):
-            self.print_red(
-                "    {number}) {exception_class}: {message}".format(
-                    number=number + 1,
-                    exception_class=exception.__class__.__name__,
-                    message="\n    ".join(str(exception).split("\n")),
-                )
-            )
-            for path, line, function_name, text in traceback.extract_tb(
-                exception.__traceback__
-            ):
-                if not self.show_testslide_stack_trace and path.startswith(
-                    os.path.dirname(__file__)
-                ):
-                    continue
-                if self.trim_path_prefix:
-                    split = path.split(self.trim_path_prefix)
-                    if len(split) == 2 and not split[0]:
-                        path = split[1]
-                self.print_cyan(
-                    '      File "{}", line {}, in {}\n'
-                    "        {}".format(path, line, function_name, text)
-                )
 
     def finish(self, not_executed_examples):
         super().finish(not_executed_examples)

--- a/testslide/runner.py
+++ b/testslide/runner.py
@@ -281,6 +281,17 @@ class ProgressFormatter(DSLDebugMixin, SlowImportWarningMixin, ColorFormatterMix
 
     def finish(self, not_executed_examples):
         super().finish(not_executed_examples)
+        if self.results["fail"] and not self.dsl_debug:
+            self.print_red("\nFAILED EXAMPLES:")
+            for fail in self.results["fail"]:
+                self.print_red(
+                    "\t{example}: {ex_class}: {ex_message}".format(
+                        example=fail["example"],
+                        ex_class=type(fail["exception"]).__name__,
+                        ex_message=str(fail["exception"]).split("\n")[0],
+                    )
+                )
+
         print("")
 
 

